core/src/main/java/org/apache/iceberg/V2Metadata.java

```diff
@@ -448,6 +449,12 @@ public Object get(int pos) {
           return wrapped.equalityFieldIds();
         case 15:
           return wrapped.sortOrderId();
+        case 16:
+          if (wrapped instanceof DeleteFile) {
```

Comment:
I am not a big fan of this `instanceof` given that it is called in a tight loop. We need it as `IndexedDataFile` is used for both data and deletes. I am not sure it is reasonable to add `IndexedDeleteFile` for 3 different fields. Thoughts?

Reply:
I think I would define `referencedDataFile` on `ContentFile` and have `DataFile` always return `null`. That's how we handle `content` in v1/v2. `DataFile` returns a constant value.

Reply:
I wanted to avoid unnecessary methods in `DataFile`. The same goes for content offset and size, which will only be applicable to deletes. Let me think.

Reply:
I think it is better not to expose these method on `DataFile`:
- Avoids reliance on validation outside of this block to ensure these fields are null for data files. We had bugs before that allowed persisting equality field IDs for data files.
- Avoids unnecessary methods that can be invoked but have no meaning, which we have to explain.